api/v1alpha1/reclaimspacejob_types.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2021 The Kubernetes-CSI-Addons Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// OperationState defines the state of the operation.
+type Result string
+
+const (
+	// ResultSucceeded represents the Succeeded operation state.
+	ResultSucceeded Result = "Succeeded"
+
+	// ResultFailed represents the Failed operation state.
+	ResultFailed Result = "Failed"
+)
+
+// TargetSpec defines the targets on which the operation can be
+// performed.
+type TargetSpec struct {
+	// PVC represents PersistentVolumeClaim name.
+	PVC string `json:"pvc,omitempty"`
+}
+
+// ReclaimSpaceJobSpec defines the desired state of ReclaimSpaceJob
+type ReclaimSpaceJobSpec struct {
+	// Target represents volume target on which the operation will be
+	// performed.
+	// +kubebuilder:validation:Required
+	Target TargetSpec `json:"target"`
+
+	// BackOffLimit specifies the number of retries allowed before marking reclaim
+	// space operation as failed. If not specified, defaults to 6.
+	// +optional
+	// +kubebuilder:default:=6
+	BackoffLimit int32 `json:"backOffLimit"`
+
+	// ActiveDeadlineSeconds specifies the duration in seconds relative to the
+	// start time that the operation might be retried; value must be positive integer.
+	// If not specified, defaults to 600 seconds.
+	// +optional
+	// +kubebuilder:default:=600
+	ActiveDeadlineSeconds int64 `json:"activeDeadlineSeconds"`
+}
+
+// ReclaimSpaceJobStatus defines the observed state of ReclaimSpaceJob
+type ReclaimSpaceJobStatus struct {
+	// Result indicates the result of ReclaimSpaceJob.
+	Result Result `json:"result,omitempty"`
+
+	// Message contains any message from the ReclaimSpaceJob.
+	Message string `json:"message,omitempty"`
+
+	// ReclaimedSpace indicates the amount of space reclaimed.
+	ReclaimedSpace resource.Quantity `json:"reclaimedSpace,omitempty"`
+
+	// Conditions are the list of conditions and their status.
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
```

Comment:
what are the conditions we are going to add as we already have result in the status?

Reply:
>what are the conditions we are going to add as we already have result in the status?

`status.result` will contain the final state `success` or `failure`, 

`status.conditions` will contain a single condition 
`type: success , status: true/false, reason/message:""` which will be used to indicate intermediate failures.

Reply:
reason can be added to the message?

Reply:
> reason can be added to the message?

I see that both are required for default metav1.Condition, 
https://pkg.go.dev/k8s.io/apimachinery@v0.21.3/pkg/apis/meta/v1#Condition

```
    // reason contains a programmatic identifier indicating the reason for the condition's last transition.
    // Producers of specific condition types may define expected values and meanings for this field,
    // and whether the values are considered a guaranteed API.
    // The value should be a CamelCase string.
    // This field may not be empty.
    // +required
    // +kubebuilder:validation:Required
    // +kubebuilder:validation:MaxLength=1024
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:Pattern=`^[A-Za-z]([A-Za-z0-9_,:]*[A-Za-z0-9_])?$`
    Reason string `json:"reason" protobuf:"bytes,5,opt,name=reason"`
    // message is a human readable message indicating details about the transition.
    // This may be an empty string.
    // +required
    // +kubebuilder:validation:Required
    // +kubebuilder:validation:MaxLength=32768
    Message string `json:"message" protobuf:"bytes,6,opt,name=message"`
```
I think it will be useful to have it. 
We can use a custom condition or remove it altogether if we don't find it to be useful later on.